kernel/ipi.c

```diff
@@ -13,15 +13,47 @@ extern void z_trace_sched_ipi(void);
 #endif
 
 
-void flag_ipi(void)
+void flag_ipi(uint32_t ipi_mask)
 {
 #if defined(CONFIG_SCHED_IPI_SUPPORTED)
 	if (arch_num_cpus() > 1) {
-		_kernel.pending_ipi = true;
+		atomic_or(&_kernel.pending_ipi, (atomic_val_t)ipi_mask);
 	}
 #endif /* CONFIG_SCHED_IPI_SUPPORTED */
 }
 
+/* Create a bitmask of CPUs that need an IPI. Note: sched_spinlock is held. */
+atomic_val_t ipi_mask_create(struct k_thread *thread)
+{
+#if defined(CONFIG_IPI_OPTIMIZE)
+	uint32_t  ipi_mask = 0;
+	uint32_t  num_cpus = (uint32_t)arch_num_cpus();
+	uint32_t  id = _current_cpu->id;
+	struct k_thread *cpu_thread;
+
+	for (uint32_t i = 0; i < num_cpus; i++) {
+		if (id == i) {
+			continue;
+		}
+
+		cpu_thread = _kernel.cpus[i].current;
+		if ((cpu_thread != NULL) &&
+		    (z_sched_prio_cmp(cpu_thread, thread) < 0)) {
```

Comment:
Zephyr allows code lines [up to 100 symbols](https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style). So let's not make code less readable :)



```suggestion
		if ((cpu_thread != NULL) && (z_sched_prio_cmp(cpu_thread, thread) < 0)) {
```